src/core/lib/promise/pipe.h

```diff
@@ -104,34 +147,66 @@ class Center {
   Poll<bool> Push(T* value) {
     GPR_DEBUG_ASSERT(send_refs_ != 0);
     if (recv_refs_ == 0) return false;
-    if (has_value_) return on_empty_.pending();
-    has_value_ = true;
+    if (value_state_ != ValueState::kEmpty) return on_empty_.pending();
+    value_state_ = ValueState::kReady;
     value_ = std::move(*value);
     on_full_.Wake();
     return true;
   }
 
+  Poll<bool> PollAck() {
+    GPR_DEBUG_ASSERT(send_refs_ != 0);
+    if (recv_refs_ == 0) return value_state_ == ValueState::kAcked;
+    if (value_state_ != ValueState::kAcked) return on_empty_.pending();
+    value_state_ = ValueState::kEmpty;
+    return true;
+  }
+
   // Try to receive a value from the pipe.
   // Return Pending if there is no value.
   // Return the value if one was retrieved.
   // Return nullopt if the send end is closed and no value had been pushed.
-  Poll<absl::optional<T>> Next() {
+  Poll<NextResult<T>> Next() {
     GPR_DEBUG_ASSERT(recv_refs_ != 0);
-    if (!has_value_) {
-      if (send_refs_ == 0) return absl::nullopt;
+    if (value_state_ != ValueState::kReady) {
+      if (send_refs_ == 0) return NextResult<T>(nullptr);
       return on_full_.pending();
     }
-    has_value_ = false;
+    return NextResult<T>(RefRecv());
+  }
+
+  void AckNext() {
+    GPR_DEBUG_ASSERT(value_state_ == ValueState::kReady);
+    value_state_ = ValueState::kAcked;
     on_empty_.Wake();
-    return std::move(value_);
+    UnrefRecv();
   }
 
+  T& value() { return value_; }
+  const T& value() const { return value_; }
+
  private:
   void ResetValue() {
     // Fancy dance to move out of value in the off chance that we reclaim some
     // memory earlier.
     [](T) {}(std::move(value_));
-    has_value_ = false;
+    value_state_ = ValueState::kEmpty;
+  }
+  enum class ValueState : uint8_t {
```

Comment:
Please document what this represents and what each of the enum values mean.

Reply:
done